### ISSUE TITLE
getXrefList(...) method is broken

### ISSUE THREAD

**Issue by OliveIsLazy** [wikipathways/rwikipathways#30]
For a few months, I've struggled with getting this method to work (first issues relating to getting geneset names that later resolved itself, and now issues with getting the actual genes in the genesets), which is strange because the issues started after the last commit.

The code I use is:

```R
# Get metadata.
pw <- listPathways("Mus musculus")

# Compile pathways.
gs <- list()
for (i in pw$id)
{
  print(i)
  gs[[length(gs)+1]] <- getXrefList(i, "L")
  print(gs[[length(gs)]])
}
```

This code used to work for me, but now it produces incomplete returns like this (I only copied the first few outputs):

```R
[1] "WP1"
[1] ""
[1] "WP10"
 [1] "11651"  "12702"  "14784"  "16186"  "16198"  "16199"  "16367"  "16451"  "16453"  "18708"  "19247"  "20416"  "20846" 
[14] "20848"  "20850"  "20851"  "26395"  "26396"  "26413"  "26417"  "269523" "384783" "54721"  "81601" 
[1] "WP103"
[1] ""
[1] "WP108"
[1] ""
[1] "WP113"
[1] ""
[1] "WP116"
[1] ""
[1] "WP1241"
[1] ""
[1] "WP1242"
[1] ""
[1] "WP1243"
[1] ""
[1] "WP1244"
[1] ""
```

I was wondering if it had to do with calling bridgedb's datasources repo, as their datasources.tsv file has been moved from the path described in location in getXrefList(...)'s help page.

Please tell me if you need more information!

**Comment by egonw** [comment]
Hi @OliveIsLazy, we're running into a few issues ourselves. Thank you for our report. I hope we can fix things as soon as possible.

**Comment by mkutmon** [comment]
@OliveIsLazy as an alternative, you can download the mouse pathway collection in GMT format which has all the entrez gene ids for every approved mouse pathway. should also be a lot faster.

```R
library(rWikiPathways)

wp.mm <- rWikiPathways::downloadPathwayArchive(date="20230210", organism="Mus musculus", format="gmt")
wp2gene <- readPathwayGMT(wp.mm)
```

**Comment by OliveIsLazy** [comment]
@mkutmon This method has helped! The dataframe had a lot of the information I needed. None of the pathways have no genes in them, though pathway WP4627 (" Lipids measured in liver metastasis from breast cancer") has only one gene it in. Here is the code block I used in case anyone else is interested. 

```R
#' Generate a list of lists, where each sublist is a WikiPathway genesets.
#' The sublist names correspond to the genesets' name.
#' The sublist values correspond to the genes in the geneset.
#' bug report: https://github.com/wikipathways/rWikiPathways/issues/30#issuecomment-1460013132
#' #' DATE: 08-03-2023
get_wikipathways_two <- function() {
  
  # Get metadata.
  wp.mm <- rWikiPathways::downloadPathwayArchive(date="20230210", organism="Mus musculus", format="gmt")
  wp2gene <- readPathwayGMT(wp.mm)
  
  # Set up key variables.
  pwid <- sort(unique(wp2gene$wpid))
  gs <- list()
  
  # For each pathway id, find the rows with a matching ID and save the associated genes in a list
  for (i in pwid)
  {
    gs[[length(gs)+1]] <- wp2gene[(wp2gene$wpid==i), "gene"]
    
    # Acts as a fail-safe in case there's an issue with the id-pathwayname mapping.
    names(gs)[length(gs)] <- unique(wp2gene[(wp2gene$wpid==i), "name"])
  }
 
  return(gs)
}
```

By the way, I checked the number of separate genesets in the function you sent me and it totals to 199, which is less than the amount I recorded from December 14th 2022, 238. And indeed, the function I posted initially still generate 238 genesets. I checked and the 199 current pathways appear to all be a subset of the previous 238. Do you know what happened to these missing pathways or how I could get access to them?

On that note, I'd also like to know if any of the pathway definitions (aka the genes in them) has changed since this date, if that's something you're able to tell me.


Thank you very much for your help and your quick response, it is much appreciated.

**Comment by mkutmon** [comment]
Ah, the GMT file only contains approved pathways. Pathways under development and not finished pathways are not included in our monthly releases. I will double-check if that is indeed the case for those 39 mouse pathways.
Regarding the changes of the pathways > the GMT files are created monthly always on the 10th. So you simply change the date in the downloadPathwayArchive function and go back to see if anything changed.

Hope that helps!

**Comment by OliveIsLazy** [comment]
I tried messing around a little and I haven't found any changes Dec 2022 and Feb 2023, but I really appreciate being able to confirm that!

This next part doesn't really relate to the coding backend, so I apologise if it's a bit rude, but I went and checked for which pathways I was missing for my analysis I'm in need of these five pathways:
- "SRF and miRs in Smooth Muscle Differentiation and Proliferation"
- "Toll-like receptor signaling pathway"                           
- "Adar1 editing defficiency immune response"                      
- "Elongation of (very) long chain fatty acids"                    
- "GPCRs, non-odorant"    

I hope to hear back soon on whether or not their gene lists can be made available again.

**Comment by DeniseSl22** [comment]
> * "SRF and miRs in Smooth Muscle Differentiation and Proliferation" 

[WP2081](https://classic.wikipathways.org/index.php/Pathway:WP2081) --> not Approved; Homology Converted

> * "Toll-like receptor signaling pathway" 

[WP1271](https://classic.wikipathways.org/index.php/Pathway:WP1271) --> Not Approved; Homology Converted

> * "Adar1 editing defficiency immune response"

 [WP3415](https://classic.wikipathways.org/index.php/Pathway:WP3415) --> Not Approved

> * "Elongation of (very) long chain fatty acids" 

[WP4491](https://classic.wikipathways.org/index.php/Pathway:WP4491) --> Not Approved

> * "GPCRs, non-odorant" 

[WP1396](https://classic.wikipathways.org/index.php/Pathway:WP1396) --> Approved, so this one should be available in the GMT download @mkutmon ...? I checked, and the GMT release of Feb. 2023 includes this PW ID, so not sure why it's not available in R?

And, how can someone make a GMT file for unapproved PWs? @mkutmon @egonw ? Maybe we should make that option available in the future?

**Comment by mkutmon** [comment]
WP1396 is also in the R table. 
Creating a GMT file for all pathways might be possible but it's likely not something we want to distribute since the pathways might not be ready for analysis yet. I restarted our indexer but the xreflist function still shows a bit of inconsistent behavior. I'll need to do more testing.

**Comment by AlexanderPico** [comment]
Agreed. GMTs serve a specific purpose for enrichment analysis. We really don't want to make or share GMTs with unapproved content.

**Comment by OliveIsLazy** [comment]
I can confirm that WP1396 is in the R table. I checked using the pathway names where the December 2022 version is called  "GPCRs, non-odorant" and the February 2023 version is of the same name except with a space added at the end, hence the mismatch.

My apologies for the confusion.

**Comment by egonw** [comment]
Okay, we worked hard to solve a number of issues, including fixes in the BridgeDb Webservice, the ID mappings databases, uptime of the data.wikipathways.org server, and the indexer. I guess one of the last two was underlying your observation. @OliveIsLazy, can you confirm your original code is working again? I get:

```R
> getXrefList("WP3415", "L")
 [1] "100038882" "100039796" "100702"    "106759"    "110454"    "12675"    
 [7] "142980"    "15945"     "15957"     "15959"     "15962"     "15975"    
[13] "15977"     "15979"     "16150"     "16151"     "17067"     "17858"    
[19] "18033"     "18035"     "19765"     "20304"     "20558"     "20684"    
[25] "21353"     "21822"     "22031"     "22034"     "223672"    "228607"   
[31] "230073"    "231655"    "23960"     "23961"     "23962"     "240327"   
[37] "240328"    "24110"     "246727"    "246728"    "246729"    "246730"   
[43] "26409"     "270151"    "320377"    "327959"    "434341"    "54123"    
[49] "54131"     "55932"     "56066"     "56417"     "56480"     "56489"    
[55] "58185"     "58203"     "60440"     "620913"    "623121"    "626578"   
[61] "66513"     "66724"     "667370"    "667373"    "67775"     "68652"    
[67] "71586"     "71898"     "74498"     "80861"     "99899"    
```

But I cannot say the results are very stable. It seems that ID mapping does not take place:

```R
> getXrefList("WP2338", "En")
[1] "ENSG00000100697" "ENSG00000113360" "ENSG00000124571" "ENSG00000128191"
> getXrefList("WP2338", "L")
[1] "5901"
```

The returned Ensembl and NCBI identifiers here are only those that are explicitly in the pathways and none of the mapped identifiers. We see this problem also on BioConductor: https://bioconductor.org/checkResults/release/bioc-LATEST/rWikiPathways/lconway-checksrc.html

Basically, it seem the reboot of the Indexer solved some problems, but not all. The Indexer is not using the latest ID mapping file, which we hope to get from @tabbassidaloii soon.

**Comment by mkutmon** [comment]
I created an issue in the indexer repo: https://github.com/wikipathways/lucene-indexer/issues/4. It has nothing to do with the new mapping databases so I need more time to actually look at the code and check what is happening and why the ID mapping doesn't work anymore.

**Comment by OliveIsLazy** [comment]
I ran the code I originally posted, and all the genesets that I needed to match did based on the name (except the "GPCRs, non-odorant" but only due to the aforementioned naming difference). It generated 238 pathways, as it did in December. I tried checking the code again against another older file I had that saved 195 of the genesets' WikiPathway IDs, and all 195 genesets were accounted for.

As @egonw pointed out, there are issues with the ID mapping. 71 of the genesets have either no or one EntrezIDs in them. Here's two examples:
```
$`Amino acid conjugation of benzoic acid`
[1] "60525"

$Apoptosis
[1] ""
```

Thank you all very much for helping me, I really appreciate the hard work.

**Comment by DeniseSl22** [comment]
@OliveIsLazy ; I'm assuming you are running your code for the mouse species, right?

- Amino acid conjugation of benzoic acid (Mm) [WP1252](https://classic.wikipathways.org/index.php/Pathway:WP1252) :
    - Website (showing 2 IDs):
![image](https://user-images.githubusercontent.com/26277832/224959134-2d10b723-d6dd-4c3e-a8b1-4bbbd2e610de.png)
    - Webservice (showing 1 ID, filtering for NCBI gene/ Entrez ID (System code L):)
![image](https://user-images.githubusercontent.com/26277832/224959684-41020ef0-3206-4764-991f-9edabc2108f8.png)
    - Webservice  (showing the other ID, filtering for Ensembl ID (System code En):)
![image](https://user-images.githubusercontent.com/26277832/224960610-ea6b94d0-9d7c-43aa-90b8-8ceabd14fac8.png)


- Apoptosis (Mm) [WP1254](https://classic.wikipathways.org/index.php/Pathway:WP1254) is showing the same behaviour (Filtering for Ensembl gives me all annotated DataNodes, filtering for Entrez Gene gives me no results at all); all DataNodes are annotated with an Ensembl ID, and the webservice XRef mapping doesn't seem to work...

@egonw and @mkutmon : seems like the mapping to other databases is (still) not happening.
When I execute the regular xRef mapping API call in the BridgeDb webservice, I do get a repsonse:
```
curl -X GET "https://webservice.bridgedb.org/Mus%20musculus/xrefs/En/ENSMUSG00000000817?dataSource=L" -H "accept: */*"

14103	Entrez Gene
```

When I try out the batch mapping, I get no matches:
```
curl -X POST "https://webservice.bridgedb.org/Mouse/xrefsBatch/En?dataSource=L" -H "accept: */*" -H "Content-Type: text/html" -d "ENSMUSG00000000817ENSMUSG00000001729ENSMUSG00000003184ENSMUSG00000003873"

TypeError: Failed to fetch
```

The same happens when I use the batch mapping, but just input 1 ID.

**Comment by OliveIsLazy** [comment]
Yes, all the code I've ran has been only for the mice pathways.

**Comment by tbrittoborges** [comment]
> @OliveIsLazy as an alternative, you can download the mouse pathway collection in GMT format which has all the entrez gene ids for every approved mouse pathway. should also be a lot faster.
> 
> ```r
> library(rWikiPathways)
> 
> wp.mm <- rWikiPathways::downloadPathwayArchive(date="20230210", organism="Mus musculus", format="gmt")
> wp2gene <- readPathwayGMT(wp.mm)
> ```

I would like to report that this stopped working recently. It seems the folder organization of the archive changed.
Should I open a new issue?

<details>
  <summary>Session info</summary>

```
R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 12 (bookworm)

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.21.so;  LAPACK version 3.11.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/Berlin
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] org.Mm.eg.db_3.17.0  AnnotationDbi_1.62.2 IRanges_2.34.1      
 [4] S4Vectors_0.38.2     Biobase_2.60.0       BiocGenerics_0.46.0 
 [7] fgsea_1.26.0         readr_2.1.4          ggrepel_0.9.4       
[10] treemapify_2.5.6     tidylog_1.0.2        stringr_1.5.1       
[13] ggplot2_3.4.4        dplyr_1.1.3         

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      jsonlite_1.8.7          magrittr_2.0.3         
  [4] farver_2.1.1            fs_1.6.3                zlibbioc_1.46.0        
  [7] ragg_1.2.6              vctrs_0.6.4             memoise_2.0.1          
 [10] Cairo_1.6-1             RCurl_1.98-1.13         ggtree_3.8.2           
 [13] base64enc_0.1-3         htmltools_0.5.7         curl_5.1.0             
 [16] gridGraphics_0.5-1      plyr_1.8.9              cachem_1.0.8           
 [19] uuid_1.1-1              ggfittext_0.10.1        igraph_1.5.1           
 [22] lifecycle_1.0.4         pkgconfig_2.0.3         gson_0.1.0             
 [25] Matrix_1.6-3            R6_2.5.1                fastmap_1.1.1          
 [28] GenomeInfoDbData_1.2.10 digest_0.6.33           aplot_0.2.2            
 [31] enrichplot_1.20.3       colorspace_2.1-0        patchwork_1.1.3        
 [34] textshaping_0.3.7       RSQLite_2.3.3           labeling_0.4.3         
 [37] fansi_1.0.5             httr_1.4.7              polyclip_1.10-6        
 [40] compiler_4.3.1          bit64_4.0.5             withr_2.5.2            
 [43] downloader_0.4          BiocParallel_1.34.2     viridis_0.6.4          
 [46] DBI_1.1.3               ggforce_0.4.1           MASS_7.3-60            
 [49] rjson_0.2.21            HDO.db_0.99.1           tools_4.3.1            
 [52] ape_5.7-1               scatterpie_0.2.1        glue_1.6.2             
 [55] nlme_3.1-163            GOSemSim_2.26.1         shadowtext_0.1.2       
 [58] grid_4.3.1              pbdZMQ_0.3-10           reshape2_1.4.4         
 [61] generics_0.1.3          gtable_0.3.4            tzdb_0.4.0             
 [64] tidyr_1.3.0             data.table_1.14.8       hms_1.1.3              
 [67] tidygraph_1.2.3         utf8_1.2.4              XVector_0.40.0         
 [70] pillar_1.9.0            yulab.utils_0.1.0       IRdisplay_1.1          
 [73] splines_4.3.1           tweenr_2.0.2            treeio_1.24.3          
 [76] lattice_0.22-5          bit_4.0.5               tidyselect_1.2.0       
 [79] GO.db_3.17.0            Biostrings_2.68.1       gridExtra_2.3          
 [82] graphlayouts_1.0.2      stringi_1.8.1           lazyeval_0.2.2         
 [85] ggfun_0.1.3             evaluate_0.23           codetools_0.2-19       
 [88] rWikiPathways_1.20.0    ggraph_2.1.0            tibble_3.2.1           
 [91] qvalue_2.32.0           ggplotify_0.1.2         cli_3.6.1              
 [94] IRkernel_1.3.2          systemfonts_1.0.5       repr_1.1.6             
 [97] munsell_0.5.0           Rcpp_1.0.11             GenomeInfoDb_1.36.4    
[100] png_0.1-8               XML_3.99-0.15           parallel_4.3.1         
[103] blob_1.2.4              clusterProfiler_4.8.3   DOSE_3.26.2            
[106] bitops_1.0-7            viridisLite_0.4.2       tidytree_0.4.5         
[109] scales_1.2.1            purrr_1.0.2             crayon_1.5.2           
[112] clisymbols_1.2.0        rlang_1.1.2             cowplot_1.1.1          
[115] fastmatch_1.1-4         KEGGREST_1.40.1  

```
</details>
